neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/bolt/impl/BoltNeo4jDatastoreProvider.java

```diff
@@ -52,17 +58,42 @@ public TransactionCoordinatorBuilder getTransactionCoordinatorBuilder(Transactio
 
 	@Override
 	public void start() {
-		if ( client == null ) {
-			try {
-				this.client = new BoltNeo4jClient( getDatabaseIdentifier(), configuration );
-				this.sequenceGenerator = new BoltNeo4jSequenceGenerator( client, getSequenceCacheMaxSize() );
-			}
-			catch (HibernateException e) {
-				// Wrap HibernateException in a ServiceException to make the stack trace more friendly
-				// Otherwise a generic unable to request service is thrown
-				throw log.unableToStartDatastoreProvider( e );
-			}
+		if ( client != null ) {
+			return;
+		}
+
+		if ( configuration.getNativeClientResource() == null ) {
+			createClient();
 		}
+		else {
+			lookupClient();
+		}
+		this.sequenceGenerator = new BoltNeo4jSequenceGenerator( client, getSequenceCacheMaxSize() );
+	}
+
+	private void createClient() {
+		try {
+			this.client = new BoltNeo4jClient( getDatabaseIdentifier(), configuration );
+		}
+		catch (HibernateException e) {
+			// Wrap HibernateException in a ServiceException to make the stack trace more friendly
+			// Otherwise a generic unable to request service is thrown
+			throw log.unableToStartDatastoreProvider( e );
+		}
+	}
+
+	private void lookupClient() {
+		try {
+			log.tracef( "Retrieving MongoDatabase from JNDI at %1$s", configuration.getNativeClientResource() );
+			Driver driver = (Driver) jndiService.locate( configuration.getNativeClientResource() );
+			this.client = new BoltNeo4jClient( driver );
+		}
+		catch (RuntimeException e) {
+			throw log.errorOnFetchJndiClientProperty( configuration.getNativeClientResource() );
+		}
+
+		// clear resources
+		this.jndiService = null;
```

Comment:
Not critical but I find it counter intuitive that the lookup method also does the clear, can we move it at the end of the start method?

Reply:
Sure I can